City-of-Helsinki/drupal-hdbt
# UHF-4235: Whitelist internal links

# UHF-4235 Whitelist internal links

## What was done
* Created a new twig template for all links on site
* Add support for external link detecting with whitelist-principle
* Add support for `tel:` and `mailto:` links
* Converted old `is_external` functionality to use new whitelist logic
* Converted `<a href...` style codes in twig templates to `link(link_title, ...` function to get whitelist support working
* Moved custom icon to left side of button label
* Other fixes
  * Unpublished -indicator z-index bug (now goes above hero image too)
  * Fix black background on main menu active-trail (when current page is on third level on main menu and you hover main level dropdown, second level no longer turns black)
  * Add bunch of small animations to arrows in links
  * Removed `target="_blank"` link attribute in places where they're not acceptable (a11y) or reasonable (internal page)
  * Cleaned up twig folder structure a tiny bit
  * Refactored content and list of links cards to use accessible link style (link around heading with pseudo element to trigger hover effects on whole card)
  * Removed duplicate code from different teasers --> node--teaser.html.twig
  * Removed duplicate code from pager twig and scss files. Now uses single source of truth for pager button.

### What was done in to [HDBT Admin theme](https://github.com/City-of-Helsinki/drupal-hdbt-admin/pull/109)
* Added consent checkbox for CKeditor links and Link target field widgets (Link fields), when trying to apply the `Opens in new tab` functionality.
* Removed `hdbt-icon--link-external` class from CKeditor link buttons.
* Removed redundant preprocess for Site settings default icon
   * Fixed Site settings default icon visibility 
* Refactored CKEditor link plugin
   * Fixed logic when switching attributes for the link
   * Sanitised unneeded false values from the plugin
   * Fixed design related icon issues when the default style is used.
   * Use Helfi api base InternalDomainResolver to parse and resolve the CKEditor links.
   * Fixed icon visibility when adding button with a design and icon removal if user decides to remove the icon.
   * Fixed automatic protocol insertion when user doesn't choose a protocol for urls.
* Select2 Icon library changes
   * Override Select2 attached library as attaching the original library will create race condition if it's included in entity form and dialog (iframe) on the same page. 
   * Fixed Select2 icon disappearing icons when adding new content and loading icons in a dialog iframe.
   * Fixed library inclusion for design selection and paragraph selection and propagating event in imagePreviewer.

### What was done in to [Helfi Platform config](https://github.com/City-of-Helsinki/drupal-helfi-platform-config/pull/263)
* Added patch to strip whitespaces from debug messages written in to markup when rendering template. 
   * Only issue is messy markup, but browser will render line breaks correctly anyway.
* Added link() twig function support to breadcrumbs.
* Added configuration update for adding 'selected-icon' and 'is-external' data attributes for CKEditor anchor tags.
* Modified Chart iframe and React 'n Share templates to use link() in twig templates.

## How to test

* Get this branch
* Get `composer require drupal/helfi_platform_config:dev-UHF-4235_whitelist_internal_links` 
* Get `composer require drupal/hdbt_admin:dev-UHF-4235_whitelist_internal_links`
* `drush cr; drush cim -y; drush updb -y; drush cr`
* Run updates and clear caches, `make drush-updb drush-cr`
* Add bunch of external, mailto and tel links to all places imaginable.
  * All navigation areas (main, top, footer, branding, dropdown, sidebar, mobile)
  * ckeditor links (default, primary, secondary, supplementary)
  * Paragraph links (list of links, content cards, banner, etc)
  * Hero buttons
  * News page links
* Check that the links work as planned

## Other PRs related to the task
* https://github.com/City-of-Helsinki/drupal-hdbt-admin/pull/109
* https://github.com/City-of-Helsinki/drupal-helfi-platform-config/pull/263
* https://github.com/City-of-Helsinki/drupal-module-helfi-api-base/pull/56
